env/io_posix.cc

```diff
@@ -409,13 +370,135 @@ size_t PosixHelper::GetUniqueIdFromFile(int fd, char* id, size_t max_size) {
   return static_cast<size_t>(rid - id);
 }
 #endif
+
+#ifdef OS_LINUX
+void LogicalBufferSizeCache::RefAndCacheLogicalBufferSize(
+    const std::set<std::pair<std::string, int>>& directories) {
+  std::map<int, size_t> fd_sizes;
+  {
+    ReadLock lock(&cache_mutex_);
+    for (auto& fname_fd : directories) {
+      if (cache_.find(fname_fd.first) == cache_.end()) {
+        fd_sizes.emplace(fname_fd.second, 0);
+      }
+    }
+  }
+  for (auto& fd_size : fd_sizes) {
+    fd_size.second = get_logical_buffer_size_(fd_size.first);
+  }
+  WriteLock lock(&cache_mutex_);
+  for (auto& fname_fd : directories) {
+    auto& v = cache_[fname_fd.first];
+    v.ref++;
+    auto fd_size = fd_sizes.find(fname_fd.second);
+    if (fd_size != fd_sizes.end()) {
+      v.size = fd_size->second;
+    }
+  }
+}
+
+void LogicalBufferSizeCache::UnrefAndTryRemoveCachedLogicalBufferSize(
+    const std::set<std::string>& directories) {
+  WriteLock lock(&cache_mutex_);
+  for (auto& dir : directories) {
+    auto it = cache_.find(dir);
+    if (it != cache_.end() && !(--it->second.ref)) {
+      cache_.erase(it);
+    }
+  }
+}
+
+size_t LogicalBufferSizeCache::GetLogicalBufferSize(const std::string& fname,
+                                                    int fd) {
+  std::string dir = fname.substr(0, fname.find_last_of("/"));
+  {
+    ReadLock lock(&cache_mutex_);
```

Comment:
I think we can even get rid of the read lock if we maintain multiple versions of the map, and always read from a consistent snapshot, kind of like mvcc. But since cache is read heavy, I think the read lock here is fine. I'll post a benchmark result from db_bench later.

Reply:
That might be too hard to maintain. I hope performacne with this ReadLock is enough.